### PR TITLE
allow share-activity to add actions (for whiteboard)

### LINKED ISSUE #269: Add support for sharing whiteboard via an activity share.
At minimum, we need to write tests to verify that sharing a whiteboard is possible and will not be broken in future versions.  The whiteboard will add an `actions` property to the files items.

I cannot label or assign, so please assign to me.

### PR DESCRIPTION
**Pull request by glhewett**
The spec for the new activity share additions to support  the whiteboard share can be found [at this link](https://sqbu-github.cisco.com/ghewett/whiteboard-integration-notes/blob/master/board_activity%20V2.md).

The idea of this PR was to change the share.addFile to accept a second `options` param that would include the actions array.  This was tested indirectly through the creation of a new share from an object or array that contained files.

The `contentCategory` of any activities that includes a whiteboard should be `documents` even though the share will look almost identical to images with the exception of the added actions array.

Both legacy and upcoming/modular SDKs have been updated.

closes #269 

CC @saurjai3 @tran2

### CODE REVIEW

File: packages/plugin-conversation/src/share-activity.js [hunk 1]
```diff
@@ -75,23 +75,25 @@ const ShareActivity = SparkPlugin.extend({
...
+  add(file, options) {
     let upload = this.uploads.get(file);
     if (upload) {
       return upload[PROMISE_SYMBOL];
     }
 
     const emitter = new EventEmitter();
 
-    upload = {
+    upload = Object.assign({
```
Comment (tran2) [inline packages/plugin-conversation/src/share-activity.js]: no lodash assign?

File: packages/plugin-conversation/src/share-activity.js [hunk 2]
```diff
@@ -316,7 +334,14 @@ ShareActivity.create = function create(conversation, object, spark) {
   });
 
   if (files) {
-    files.forEach((file) => share.add(file));
+    files.forEach((file) => {
+      if (Array.isArray(file)) {
```
Comment (tran2) [inline packages/plugin-conversation/src/share-activity.js]: when does this happen?

File: packages/plugin-conversation/src/share-activity.js (hunk 1 above)
Comment (ianwremmel) [inline packages/plugin-conversation/src/share-activity.js]: yep. Object.assign is the way to go. Most (though not all) lodash methods that operate on collections now exist in the core language in some way.

map, reduce, filter, includes, find, findIndex are all part of array now.

File: packages/plugin-conversation/src/share-activity.js (hunk 2 above)
Comment (glhewett) [inline packages/plugin-conversation/src/share-activity.js]: This happens when and activity share is created from an object.

`spark.conversation.share(conversation, { files: [....] }`

File: packages/plugin-conversation/src/share-activity.js (hunk 1 above)
Comment (glhewett) [inline packages/plugin-conversation/src/share-activity.js]: I may need to make some modifications to this PR. To this PR because I use `lodash.map` instead of the `Array.map`.

File: packages/plugin-conversation/package.json [hunk 3]
```diff
@@ -25,6 +25,7 @@
     "@ciscospark/test-helper-file": "^0.7.0-alpha",
     "@ciscospark/test-helper-mocha": "^0.7.0-alpha",
     "@ciscospark/test-helper-make-local-url": "^0.7.0-alpha",
+    "@ciscospark/test-helper-sinon": "^0.7.0-alpha",
```
Comment (glhewett) [inline packages/plugin-conversation/package.json]: @ianwremmel, I noticed that this was not included, which was strange.  I found this by cleaning all the modules, and running bootstrap. 😃 I could only smile in hindsight to that one.

File: packages/plugin-conversation/src/share-activity.js (hunk 2 above)
Comment (glhewett) [inline packages/plugin-conversation/src/share-activity.js]: I am unsure what situation you are referring to that would fail, @ianwremmel.  My example might be from legacy `spark.conversation.share`.  Looks like the new one will take an object or an array of files.

File: packages/plugin-conversation/src/share-activity.js [hunk 4]
```diff
@@ -316,7 +334,14 @@ ShareActivity.create = function create(conversation, object, spark) {
...
   if (files) {
-    files.forEach((file) => share.add(file));
+    files.forEach((file) => {
+      if (isArray(file)) {
+        Reflect.apply(share.add, share, file);
+      }
+      else {
+        share.add(file);
+      }
+    });
```
Comment (glhewett) [inline packages/plugin-conversation/src/share-activity.js]: I may need to add a comment.  In this case file will be an array.  file[0] is the file, and file[1] are the options.

**Comment by glhewett** [comment]
@ianwremmel, when creating the PR, I thought about how I wanted to use the API, and not how the web-client was going to use it. What I needed to be able to do was not break existing code while adding the ability to pass more options into the `addFile` function of the share activity.  In doing so, I created a second option in the `addFile` function called options.  The options will include property called actions, which has the board information in it.  The board share is almost exactly the same as an image share (the image is a rasterized version of the board) except that it adds an actions property.

Briefly, I tried to work through the process of sneaking the actions property into the file buffer, but I could not add the property into the file safely.  When I got aggressive (assign), things did not work well. 

Today, I am trying to integration the functionality into the web-client, and I have been met with a bit of complexity.  The strategy that I was going to use was to change the file param to an array (`[file, options]`) in order to provide the ability for the actions property to live through out the life of the object until stored on the server.

The way that the files are added are just like images.  They are selected from a board listing screen (new), and they are added to the StagedFile until they are finally submitted to the conversation.  There seem to be resizing and sanitizing that is going on that I would like the whiteboard share to take advantage of, so it looks like a bit of a complicated change.  I am going to follow the path of addFile throughout the web-client and add a conditional like so...

```
file = file
options = {}

if (isArray(file)) {
  file = file[0]
  options = options[1]
}
```

Before I made such an ugly change, I want to bounce it off of you and @saurjai3.

Do either of you have any ideas on how I might make this change a bit more straight forward?  Do you agree with my approach to this PR?  Maybe the web-client changes won't be that bad.

**Comment by glhewett** [comment]
Possibly converting the ArrayBuffer on create to a File or Blob may provide a place to hide the actions property.  I am currently working in the SquaredAdapter to see if I can hide the actions in the `File` or `Blob` then extract the action upon adding a new file.

**Comment by glhewett** [comment]
Doing the conversion of the ArrayBuffer may only be valid for the integration tests and the use of the fixtures.  it looks like the web client is using a Blob for the binary data, which provides the ability to add the actions property into the file/blob object.

I am not stuck, I am just soliciting feedback.  I think I have a direction that I am going to move in tomorrow.  None the less, I am removing the "ready for review" label for now.

**Comment by ianwremmel** [comment]
@glhewett, i didn't entirely follow you're first comment, but I'm against passing disparate parameters as positional arguments in an array (I actually try to avoid relying on positional arguments in general and pass most function parameters as objects to allow parameter naming). 

While the web client relies pretty much entirely on Blobs, we should aim to support arraybuffers for Buffer support on the node side of things.

File: src/client/services/conversation/share-activity/share-activity-base.js [hunk 5]
```diff
@@ -118,9 +121,24 @@ var ShareActivityBase = SparkBase.extend(
    * @returns {Promise} Resolves with the updated activity
    */
   addFile: function addFile(file) {
+    /* eslint complexity: [0] */
```
Comment (glhewett) [inline src/client/services/conversation/share-activity/share-activity-base.js]: @ianwremmel I just wanted to acknowledge this.  You think this is because I was the last to add another path to the flow?

File: src/client/services/conversation/share-activity/share-activity-base.js [hunk 6]
```diff
@@ -404,12 +423,27 @@ var ShareActivityBase = SparkBase.extend(
   },
 
   /**
+   * @param {Array} items
+   * @param {string} mimeType
+   * @private
+   * @returns {boolean}
+   */
+  _itemContainsActionWithMimeType: function _itemContainsActionWithMimeType(items, mimeType) {
+    return some(map(items, (item) => some(item.actions, ['mimeType', mimeType])));
```
Comment (glhewett) [inline src/client/services/conversation/share-activity/share-activity-base.js]: No arrow functions in legacy SDK!!

File: packages/plugin-conversation/src/share-activity.js [hunk 7]
```diff
@@ -78,6 +78,20 @@ const ShareActivity = SparkPlugin.extend({
    * @returns {EventEmittingPromise}
    */
   add(file) {
+    let actions;
+
+    // ArrayBuffer might be located in an object with the actions.
+    if (has(file, `file`) && has(file, `actions`)) {
```
Comment (ianwremmel) [inline packages/plugin-conversation/src/share-activity.js]: differentiating blobs and arraybuffers like this feels really brittle. why is this necessary?

File: packages/plugin-conversation/src/share-activity.js [hunk 8]
```diff
@@ -255,10 +270,26 @@ const ShareActivity = SparkPlugin.extend({
 
   /**
    * @param {Array} items
+   * @param {string} mimeType
+   * @private
+   * @returns {boolean}
+   */
+  _itemContainsActionWithMimeType(items, mimeType) {
+    return some(items.map((item) => some(item.actions, [`mimeType`, mimeType])));
```
Comment (ianwremmel) [inline packages/plugin-conversation/src/share-activity.js]: the rest of the sdk uses object instead of array notation for lodash predicates. can you changes this to `{mimeType: mimeType}`?

File: packages/plugin-conversation/test/integration/spec/share.js [hunk 9]
```diff
@@ -217,6 +217,35 @@ describe(`plugin-conversation`, function() {
               });
           }));
       });
+
+      it(`shares a whiteboard`, () => spark.conversation.share(conversation, [{
+        file: sampleImageSmallOnePng,
+        actions: [{
+          type: `edit`,
+          mimeType: `application/x-cisco-spark-whiteboard`,
```
Comment (ianwremmel) [inline packages/plugin-conversation/test/integration/spec/share.js]: Why do actions have mimeTypes?

File: packages/plugin-conversation/test/integration/spec/share.js [hunk 10]
```diff
@@ -217,6 +217,35 @@ describe(`plugin-conversation`, function() {
...
+          assert.isEncryptedActivity(activity);
+          assert.isFileItem(activity.object.files.items[0]);
+          assert.isThumbnailItem(activity.object.files.items[0].image);
+          assert.equal(activity.object.contentCategory, `documents`);
+          assert.isArray(activity.object.files.items[0].actions);
+          assert.equal(activity.object.files.items[0].actions[0].type, `edit`);
+          assert.equal(activity.object.files.items[0].actions[0].mimeType, `application/x-cisco-spark-whiteboard`);
+          assert.equal(activity.object.files.items[0].actions[0].url, `https://boards.example.com/boards/1`);
+
+          return Promise.all([
```
Comment (ianwremmel) [inline packages/plugin-conversation/test/integration/spec/share.js]: You don't need a Promise.all here, you've only got a single promise.

File: packages/plugin-conversation/test/integration/spec/share.js [hunk 11]
```diff
@@ -217,6 +217,35 @@ describe(`plugin-conversation`, function() {
...
+            spark.conversation.download(activity.object.files.items[0])
+              .then(tap((f) => assert.equal(f.type, `image/png`)))
+          ]);
+        })
+
+        .then(([file0]) => Promise.all([
```
Comment (ianwremmel) [inline packages/plugin-conversation/test/integration/spec/share.js]: Again, no need for promise.all

File: src/client/services/conversation/share-activity/share-activity-base.js (hunk 5 above)
Comment (ianwremmel) [inline src/client/services/conversation/share-activity/share-activity-base.js]: Yep, it looks like you added a least three branches to this function.

File: src/client/services/conversation/share-activity/share-activity-base.js [hunk 12]
```diff
@@ -118,9 +121,24 @@ var ShareActivityBase = SparkBase.extend(
    * @returns {Promise} Resolves with the updated activity
    */
   addFile: function addFile(file) {
+    /* eslint complexity: [0] */
+    var actions;
+
+    // ArrayBuffer might be located in an object with the actions.
+    if (has(file, 'file') && has(file, 'actions')) {
```
Comment (ianwremmel) [inline src/client/services/conversation/share-activity/share-activity-base.js]: Please see my comment on the plugin regarding blobs and arraybuffers

File: src/client/services/conversation/share-activity/share-activity-base.js [hunk 13]
```diff
@@ -404,12 +423,29 @@ var ShareActivityBase = SparkBase.extend(
...
 
   /**
+   * @param {Array} items
+   * @param {string} mimeType
+   * @private
+   * @returns {boolean}
+   */
+  _itemContainsActionWithMimeType: function _itemContainsActionWithMimeType(items, mimeType) {
+    return some(map(items, function doSomeActionsContainMimeType(item) {
+      return some(item.actions, ['mimeType', mimeType]);
```
Comment (ianwremmel) [inline src/client/services/conversation/share-activity/share-activity-base.js]: use object notation

File: src/client/services/conversation/share-activity/share-activity-base.js [hunk 14]
```diff
@@ -404,12 +423,29 @@ var ShareActivityBase = SparkBase.extend(
   },
 
   /**
+   * @param {Array} items
+   * @param {string} mimeType
+   * @private
+   * @returns {boolean}
+   */
+  _itemContainsActionWithMimeType: function _itemContainsActionWithMimeType(items, mimeType) {
+    return some(map(items, function doSomeActionsContainMimeType(item) {
```
Comment (ianwremmel) [inline src/client/services/conversation/share-activity/share-activity-base.js]: Why two `some`s here but only one in the plugin?

File: packages/plugin-conversation/src/share-activity.js (hunk 7 above)
Comment (glhewett) [inline packages/plugin-conversation/src/share-activity.js]: ArrayBuffers do not have properties.  In order to store the actions within the file, this differentiation was necessary.  There is some discussion about that earlier in this PR discussion.

It is tough to efficiently communicate here, so lets talk on spark if you have further questions.  I would love some feedback.  Personally, I have only been able to come up with mediocre solutions to how to this challenge.

File: packages/plugin-conversation/src/share-activity.js (hunk 8 above)
Comment (glhewett) [inline packages/plugin-conversation/src/share-activity.js]: You are right, I am not sure what I was thinking with that.

File: packages/plugin-conversation/test/integration/spec/share.js (hunk 9 above)
Comment (glhewett) [inline packages/plugin-conversation/test/integration/spec/share.js]: That was part of the design put together by the team.  The mime type is support to specify the application that is used to open the file.   When a video is available for the whiteboard, we will use a video mimeType and a URL to the stream.  (Not all clients will consume that)

Comment (glhewett) [inline packages/plugin-conversation/test/integration/spec/share.js]: https://sqbu-github.cisco.com/ghewett/whiteboard-integration-notes/blob/master/board_activity%20V2.md

File: src/client/services/conversation/share-activity/share-activity-base.js (hunk 14 above)
Comment (glhewett) [inline src/client/services/conversation/share-activity/share-activity-base.js]: There are two, but it looks a bit different without the arrow functions.  The arrow functions broke the web-client. :smile:

File: packages/plugin-conversation/src/share-activity.js (hunk 7 above)
Comment (ianwremmel) [inline packages/plugin-conversation/src/share-activity.js]: ok, i guess my point is, rather than have two different behaviors, why not just make blobs identical to arraybuffers?

Comment (glhewett) [inline packages/plugin-conversation/src/share-activity.js]: That is a great question.  My view was that making an exception was going to be safer than changing the API.  The consequences of the change the API would be big.  An alternative would be to add a new function called `addAsOptions({file: blob/arraybuffer, actions: {})`.  Is that closer to what you are thinking?

Comment (glhewett) [inline packages/plugin-conversation/src/share-activity.js]: I will take a look at that.

**Comment by glhewett** [comment]
Hi @ianwremmel,  I have refactored the  code to remove the complexity and brittleness comments.  What I have done is to simply add an second parameter to `add`/`addFile` called options.  These options will include information that needs to be stored with the file in the share activity.  The only option supported is obviously `actions`.  The consequence to this that we cannot provide this functionality to the `spark.conversation.share` convenience function.  We need to create the share activity, add files one by one, and then submit.  This is fine for the use in the web-client.
